src/Client/GuzzleFileFetcher.php

```diff
@@ -54,30 +56,37 @@ public static function createFromUri(string $baseUri) : self
     /**
      * {@inheritdoc}
      */
-    public function fetchFile(string $fileName, int $maxBytes):string
+    public function fetchFile(string $fileName, int $maxBytes): PromiseInterface
     {
-        try {
-            $response = $this->client->request('GET', $fileName);
-        } catch (ClientException $e) {
-            if ($e->getCode() === 404) {
-                throw new RepoFileNotFound("$fileName not found", 0, $e);
+        $onFulfilled = function (ResponseInterface $response) use ($fileName, $maxBytes) {
+            $body = $response->getBody();
+            $contents = $body->read($maxBytes);
+
+            // If we reached the end of the stream, we didn't exceed the maximum
+            // number of bytes.
+            if ($body->eof() === true) {
+                return $contents;
             } else {
-                // Re-throwing the original exception will blow away the
-                // backtrace, so wrap the exception in a more generic one to aid
-                // in debugging.
-                throw new \RuntimeException($e->getMessage(), $e->getCode(), $e);
+                throw new DownloadSizeException("$fileName exceeded $maxBytes bytes");
             }
-        }
+        };
 
-        $body = $response->getBody();
-        $contents = $body->read($maxBytes);
-        // If we reached the end of the stream, we didn't exceed the maximum
-        // number of bytes.
-        if ($body->eof() === true) {
-            return $contents;
-        } else {
-            throw new DownloadSizeException("$fileName exceeded $maxBytes bytes");
-        }
+        $onRejected = function (\Throwable $e) use ($fileName) {
+            if ($e instanceof ClientException) {
+                if ($e->getCode() === 404) {
+                    throw new RepoFileNotFound("$fileName not found", 0, $e);
+                } else {
+                    // Re-throwing the original exception will blow away the
+                    // backtrace, so wrap the exception in a more generic one to aid
+                    // in debugging.
+                    throw new \RuntimeException($e->getMessage(), $e->getCode(), $e);
+                }
+            } else {
+                throw $e;
+            }
+        };
+        return $this->client->requestAsync('GET', $fileName)
```

Comment:
For posterity, since there's a lot else going on in this PR: the essential change change from synchronous to asynchronous happens right here.

```diff
- $this->client->request('GET', $fileName)
+ $this->client->requestAsync('GET', $fileName)
```